app/models/assignment.rb

```diff
@@ -850,7 +853,67 @@ def get_num_valid
              .count(&:is_valid?)
   end
 
-  def get_num_marked(ta_id = nil)
+  def marked_result_ids_for(ta_id)
+    cache_ta_results
+    @cache_ta_results.dig(ta_id, :marked_result_ids) || []
+  end
+
+  def cache_ta_results
+    return @cache_ta_results if defined? @cache_ta_results
+
+    data = current_results.joins(grouping: :tas).pluck('tas.id', 'results.id', 'results.marking_state')
+    # Group results by TA ID
+    grouped_data = data.group_by { |ta_id, _result_id, _marking_state| ta_id }
+
+    results_by_id = current_results.includes(:marks).index_by(&:id)
+    @cache_ta_results = grouped_data.each_with_object({}) do |(ta_id, results), hash|
+      total_results = results.map { |_, result_id, _| result_id }
+
+      if self.assign_graders_to_criteria
+        # Get the list of criteria IDs assigned to this TA
+        assigned_criteria_ids = self.criteria.flat_map do |criterion|
+          criterion.criterion_ta_associations.select { |assoc| assoc.ta_id == ta_id }.map { criterion.id }
+        end
+
+        criteria_count = assigned_criteria_ids.size
+        if criteria_count == 0
+          # If the TA has no assigned criteria, fallback to using marking_state to determine marked results
+          complete_results = results.select do |_, _, marking_state|
+            marking_state ==
+                      Result::MARKING_STATES[:complete]
+          end
+          marked_result_ids = complete_results.map { |_, result_id, _| result_id }
+        else
+          # Count results where all assigned criteria have been marked
+          marked_result_ids = total_results.select do |result_id|
```

Comment:
Similar to my above comment, this part is okay but will still result in an N+1 query because of the looping over results.

Instead, use one query to get *a count of non-nil marks for the assigned criteria, and for each result*. A good starting point is something like `Result.joins(:marks).where(id: ..., mark: {criterion_id: ..., mark: ...})` and then use a [grouping operation](https://guides.rubyonrails.org/active_record_querying.html#grouping).

Note: even if you do the above, there actually will still be an N+1 query because you'll need to do this once per TA. Not ideal, but I think resolving that N+1 query will be harder so I think it's best to forgo it for this PR to get this merged in, and then I'll give it to another student to continue this work in the future.